Tests/SwiftParserTest/Assertions.swift

```diff
@@ -316,32 +309,68 @@ func AssertParse(
+  substructure expectedSubstructure: Syntax? = nil,
+  substructureAfterMarker: String = "START",
+  diagnostics expectedDiagnostics: [DiagnosticSpec] = [],
+  applyFixIts: [String]? = nil,
+  fixedSource expectedFixedSource: String? = nil,
+  file: StaticString = #file,
+  line: UInt = #line
+) {
+  AssertParse(
+    markedSource,
+    { source in
+      var parser = Parser(source)
+      return parse(&parser)
+    },
+    substructure: expectedSubstructure,
+    substructureAfterMarker: substructureAfterMarker,
+    diagnostics: expectedDiagnostics,
+    applyFixIts: applyFixIts,
+    fixedSource: expectedFixedSource,
+    file: file,
+    line: line)
 }
 
-/// Parse `markedSource` and perform the following assertions:
-///  - The parsed syntax tree should be printable back to the original source code (round-tripping)
-///  - Parsing produced the given `diagnostics` (`diagnostics = []` asserts that the parse was successful)
-///  - If `fixedSource` is not `nil`, assert that applying the Fix-Its with the
-///    messages in `applyFixIts` (or all if `applyFixIts` is `nil`) all fixes
-///    from the diagnostics produces `fixedSource`
+/// Removes any test markers from `markedSource` (1) and parses the result
+/// using `parse`. By default it only checks if the parsed syntax tree is
+/// printable back to the origin source, ie. it round trips.
+///
+/// (1) `markedSource` is source that can include markers of the form `1️⃣`,
+/// to be used as locations in the following parameters when they exist.
 ///
-/// The source file can be marked with markers of the form `1️⃣` to mark source locations that can be referred to by `diagnostics`.
-/// These markers are removed before parsing the source file.
-/// By default, `DiagnosticSpec` asserts that the diagnostics is produced at a location marked by `1️⃣`.
-/// `parseSyntax` can be used to adjust the production that should be used as the entry point to parse the source code.
-/// If `substructure` is not `nil`, asserts that the parsed syntax tree contains this substructure.
-func AssertParse<Node: RawSyntaxNodeProtocol>(
+/// - Parameters:
+///   - substructure: Asserts the parsed syntax tree contains this structure.
+///   - substructureAfterMarker: Changes the position to start the structure
+///     assertion from, ie. allows matching a particular substructure rather
+///     than the whole source.
+///   - diagnostics: Asserts the given diagnostics were output, by default it
+///     asserts the parse was successful (ie. it has no diagnostics). Note
+///     that `DiagnosticsSpec` uses the location marked by `1️⃣` by default.
+///   - applyFixIts: Applies only the fix-its with these messages.
+///   - fixedSource: Asserts that the source after applying fix-its matches
+///     this string.
+func AssertParse<S: SyntaxProtocol>(
   _ markedSource: String,
-  _ parseSyntax: (inout Parser) -> Node,
+  _ parse: (String) -> S,
   substructure expectedSubstructure: Syntax? = nil,
```

Comment:
Seeing this, I'm wondering whether we can get rid of the custom entry point here altogether. It didn't really turn out to be super useful IMO.

Reply:
Which one? I was going to change these methods to a struct + `check` or similar after this PR.

Reply:
I was thinking that `parse` here could just always be `SourceFileSyntax.parse`. I'm still not 100% sure about it though.

Reply:
Oh I see. We currently expose a few parse methods (eg. decl/statement/expr/pattern/memberdecl), so IMO it probably makes sense to keep these still.